src/core_codemods/lazy_logging.py

```diff
@@ -0,0 +1,160 @@
+            - pattern: logging.getLogger(...).$FUNC(..., ... % ..., ...)
+            {_pattern_inside}
+            {_log_funcs}
+          - patterns:
+            - pattern: $VAR.$FUNC(..., ... % ..., ...)
+            - pattern-inside: |
+                import logging
+                ...
+                $VAR = logging.getLogger(...)
+                ...
+            {_log_funcs}
+          - patterns:
+            - pattern: logging.$FUNC(..., $ANYTHING + ..., ...)
+            {_pattern_inside}
+            {_log_funcs}
+          - patterns:
+            - pattern: logging.getLogger(...).$FUNC(..., $ANYTHING + ..., ...)
+            {_pattern_inside}
+            {_log_funcs}
+          - patterns:
+            - pattern: $VAR.$FUNC(..., $ANYTHING + ..., ...)
+            - pattern-inside: |
+                import logging
+                ...
+                $VAR = logging.getLogger(...)
+                ...
+            {_log_funcs}
+        """
+
+    def on_result_found(self, original_node, updated_node):
+        match updated_node.func:
+            case cst.Name(value="log") | cst.Attribute(attr=cst.Name(value="log")):
+                # logging.log(INFO, ...), log(INFO, ...)
+                first_arg = [original_node.args[0]]
+                remaining_args = list(original_node.args[2:])
+                binop = original_node.args[1].value
+            case _:
+                first_arg = []
+                remaining_args = list(original_node.args[1:])
+                binop = original_node.args[0].value
+
+        if set(self.all_operators(binop)) == {cst.Add, cst.Modulo}:
+            # TODO: handle more complex case of str concat that uses both `%` and `+` operators
+            return updated_node
+
+        match binop.operator:
+            case cst.Modulo():
+                new_args = self.make_args_for_modulo(binop)
+            case cst.Add():
+                if (new_args := self.make_args_for_plus(binop)) is None:
+                    return updated_node
+        return updated_node.with_changes(args=first_arg + new_args + remaining_args)
+
+    def make_args_for_plus(self, binop: cst.BinaryOperation) -> Optional[list[cst.Arg]]:
+        if self.is_str_concat(binop):
+            # Do not change explicit str concat, e.g.: `logging.info("one" + "two")
+            return None
+
+        if isinstance(binop.left, cst.SimpleString) and "%" in binop.left.value:
+            # Do no change `logging.info("Something: %s " + var)` since intention is unclear
+            return None
+        left_type = infer_expression_type(self.resolve_expression(binop.left))
+        right_type = infer_expression_type(self.resolve_expression(binop.right))
+        if left_type != right_type or (type_both_sides := left_type) not in {
+            BaseType.STRING,
+            BaseType.BYTES,
+        }:
+            # Cannot concat different types.
+            # Skip logging ints, etc. Eg: `logging.info(2+2)`
+            return None
```

Comment:
You can't concat different types in Python with `+`, so if we assume the code compiles, we could avoid the first check.

Reply:
We do have to assume that it's valid python code (otherwise `libcst` won't even parse it). I guess this is more of a runtime exception case.
It's fine to go without this assumption, but `infer_expression` does assume that for the `BinaryOperation` cases. We can adjust it though.

Reply:
1. libcst does parse it:
```
  File "/Users/dani/codemodder-python/test.py", line 4, in <module>
    logging.info("hi" + 2)
                 ~~~~~^~~
TypeError: can only concatenate str (not "int") to str
```
run this: no error
```
import libcst as cst
source_code = """
import logging
logging.info("hi" + 2)
"""
module = cst.parse_module(source_code)
```

2. I reiterate, this check is needed for cases like `"Scientific Value: " + format(scientific_val, '.2e'))`